src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,321 @@
+                    }
+                }
+                else
+                {
+                    context.ReportDiagnostic(symbol.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+                }
+            }
+        }
+
+        private static void BuildSymbolInformationFromOperations(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            IOperation operation = context.Operation;
+            if (OperationUsesPreviewFeatures(operation, requiresPreviewFeaturesSymbols, out ISymbol? symbol))
+            {
+                if (operation is ICatchClauseOperation catchClauseOperation)
+                {
+                    operation = catchClauseOperation.ExceptionDeclarationOrExpression;
+                }
+
+                context.ReportDiagnostic(operation.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+            }
+        }
+
+        private static bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, [NotNullWhen(true)] out ISymbol? symbol)
+        {
+            symbol = GetOperationSymbol(operation);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            return OperationUsesPreviewFeatures(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
+                IFieldReferenceOperation fOperation => fOperation.Field,
+                IDelegateCreationOperation dOperation => dOperation.Type,
+                IEventReferenceOperation eOperation => eOperation.Member,
+                IUnaryOperation uOperation => uOperation.OperatorMethod,
+                IBinaryOperation bOperation => bOperation.OperatorMethod,
+                IArrayCreationOperation arrayCreationOperation => (arrayCreationOperation.Type as IArrayTypeSymbol)?.ElementType,
+                ICatchClauseOperation catchClauseOperation => catchClauseOperation.ExceptionType,
+                _ => null,
+            };
+
+        private static bool OperationUsesPreviewFeatures(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            var containingSymbol = symbol.ContainingSymbol;
+            // Namespaces do not have attributes
+            while (containingSymbol is INamespaceSymbol)
+            {
+                containingSymbol = containingSymbol.ContainingSymbol;
+            }
+
+            if (containingSymbol != null)
+            {
+                if (HasPreviewAttribute(containingSymbol, requiresPreviewFeaturesSymbols))
```

Comment:
Is it enough to check only one level up? If the containing symbol/method was not annotated but what if the contacting type or assembly was annotated?

Reply:
> assembly was annotated

If the current assembly is annotated, we will detect it. If dependencies are annotated, we won't currently. I'll implement that in another PR (also called it out in the PR description). The aim here is to get this PR in soon so Tanner can give me more feedback. 

> Is it enough to check only one level up?

For now this seems to work fine. I haven't come across a case where we need to look 2 levels up. If we do, then we can change it. Looking only 1 level up feels clean. Any more than that and I fear we will have holes in our detection.

Reply:
> If the current assembly is annotated, we will detect it. If dependencies are annotated, we won't currently. 

I meant what if the assembly of referenced API is annotated, not the current assembly

> The aim here is to get this PR in soon so Tanner can give me more feedback.

You can also give him the DLL of your analyzer and he can replace the analyzer wherever it is using

Reply:
> I meant what if the assembly of referenced API is annotated

Right right, that's what I meant. We don't detect it right now. That is by design because I want to get this PR in first. I'll add that detection in a follow up PR.

> You can also give him the DLL of your analyzer and he can replace the analyzer wherever it is using

We tried that 2-3 weeks back and it didn't work unfortunately. We also couldn't figure out why. I think it had something to do with using the latest compiler + sdk + analyzer bits combination :/ That's why we decided to get the first pass of this analyzer in, so that we can get some feedback while Tanner writes unit tests

Reply:
👌